man/xlog.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms
.\"and conditions for redistribution.
.\"
.\" $Id$
.TH XLOG 3X "15 June 1993"
xlog_parms, xlog_create, xlog_destroy, xlog_write, xlog_control -- general purpose logging facility
.SH SYNOPSIS
.LP
.nf
.ft B
#include "xlog.h"
.LP
.ft B
xlog_h xlog_create( type, id, flags, ... )
xlog_e type ;
char *id ;
int flags ;
.LP
.ft B
int xlog_parms( type, ... )
xlog_e type ;
.LP
.ft B
void xlog_destroy( xlog )
xlog_h xlog ;
.LP
.ft B
void xlog_write( xlog, buf, len, flags, ... )
xlog_h xlog ;
char buf[] ;
int len ;
int flags ;
.LP
.ft B
int xlog_control( xlog, cmd, ... )
xlog_h xlog ;
xlog_cmd_e cmd ;
.SH DESCRIPTION
The purpose of this library is to provide a general purpose logging facility
by providing
.I xlogs,
logging objects that may be connected to various existent logging facilities.
Currently, the only logging facilities supported are
.I "syslog(3)"
and logging to files.
Log entries are timestamped lines which may contain arbitrary information.
.\" ********************* xlog_create ***********************
.LP
.B xlog_create()
creates a new xlog of the specified
.I type.
Possible type values are:
.RS
.TP 20
.SB XLOG_SYSLOG
Varargs: \fIint facility, int priority\fP.
The xlog will be connected to
.I "syslog(3)."
.I facility
determines the syslog facility to use for logged messages and
.I priority
is the default message priority.
.TP
.SB XLOG_FILELOG
Varargs: \fIchar *pathname, int flags [, int flags]\fP.
The xlog will be connected to the file identified by
.I pathname.
The variable part of the argument list has the same semantics as the
argument list of the
.I "open(2)"
system call.
.RE
.LP
All xlogs have an id, specified by the
.I id
argument. The
.I flags
argument is formed by ORing one or more of the following constants:
.RS
.TP 20
.SB XLOG_NO_ERRNO
do not replace
.I "%m"
with an explanation of the current value of errno.
.TP
.SB XLOG_NO_SIZECHECK
.I "(XLOG_FILELOG only)"
do not perform size checks on the file.
.TP
.SB XLOG_PRINT_ID
precede each log entry with the xlog id
.TP
.SB XLOG_PRINT_PID
precede each log entry with the process id
(the process id will follow the xlog id)
.RE
.LP
Flags that do not apply to the xlog are ignored.
The contant
.SM XLOG_NOFLAGS
can be used if you don't want to specify any flags.
.\" ********************* xlog_parms ***********************
.LP
.B xlog_parms()
sets default parameters for the specified xlog
.I type:
.RS
.TP 20
.SB XLOG_SYSLOG
3 arguments are expected which correspond one-to-one to the arguments of
.I "openlog(3)."
The defaults, in case this function is not used, are:
"XLOG", LOG_PID + LOG_NOWAIT, LOG_USER
.TP
.SB XLOG_FILELOG
No action.
.RE
.LP
.B xlog_parms()
should be invoked before any xlogs of the specified type
are created.
.\" ********************* xlog_destroy ***********************
.LP
.B xlog_destroy()
destroys an xlog. The action taken depends on the type of the xlog:
.RS
.TP 20
.SB XLOG_SYSLOG
if this is the last xlog using syslog, then
.I "closelog(3)"
is invoked.
.TP
.SB XLOG_FILELOG
The file is closed.
.RE
.\" ********************* xlog_control ***********************
.LP
.B xlog_control()
applies control operations to an xlog. Certain operations are common to
all xlogs while others are type-specific. The common ones are:
.RS
.TP 15
.SB XLOG_LINK
Argument list: \fIxlog_h link_to\fP.
Link the specified xlog to the one provided as argument.
(if the argument is
.SM NULL
any existent link is severed).
Linking xlogs has the effect that if one finds an error it uses the
other to report it.
.TP
.SB XLOG_CALLBACK
Argument list: \fIvoid (*callback)(), void *arg\fP.
This function will be invoked in case of error while writing a log
entry. It will be given
3 arguments: the xlog handle, an integer that indicates the type
of error and the
.I arg
specified in this call. Possible errors include:
.RS
.TP 15
.SB XLOG_ENOMEM
lack of memory
.TP
.SB XLOG_EOPEN
.I "open(2)"
failed
.TP
.SB XLOG_EFSTAT
.I "fstat(2)"
failed
.TP
.SB XLOG_ESIZE
hard limit exceeded
.RE
.TP
.SB XLOG_SETFLAG
Argument list: \fIint flag, int *value\fP.
The value of
.I flag
(one of those listed before) is set according to
.I "*value"
which should be either 0 or 1.
The old flag value is placed in
.I "*value."
.TP
.SB XLOG_GETFLAG
Argument list: \fIint flag, int *value\fP.
The value of
.I flag
(one of those listed before) is placed in
.I "*value."
.RE
.LP
Xlogs of type
.B XLOG_SYSLOG
also support the following operations:
.RS
.TP 15
.SB XLOG_FACILITY
Argument list: \fIint facility\fP.
Sets the syslog facility to the specified value.
.TP
.SB XLOG_LEVEL
Argument list: \fIint level\fP.
Sets the default syslog level for this xlog to the specified value.
.TP
.SB XLOG_PREEXEC
Argument list: \fIvoid\fP.
Prepares the xlog for an impending exec operation
.TP
.SB XLOG_POSTEXEC
Argument list: \fIvoid\fP.
Informs the xlog that the exec failed
.RE
.LP
Xlogs of type
.B XLOG_FILELOG
also support the following operations:
.RS
.TP 15
.SB XLOG_LIMITS
Argument list: \fIunsigned soft_limit, unsigned hard_limit\fP.
Sets soft and hard limits on the size of the file.
When any of the limits is exceeded a message is sent to the linked xlog.
(if there is no linked xlog, no message is sent)
When the soft limit is exceeded a warning message is sent to the linked xlog
(if the linked xlog is of the
.SB XLOG_SYSLOG
type, the message will be sent at the
.I LOG_ALERT
level).
If the exceeded limit is the hard limit, logging is terminated.
The actual file size is checked every time this operation is applied to
the file.
If logging was terminated because the hard limit was exceeded and
this operation increases the hard limit beyond the actual file size,
logging will be resumed.
.TP
.SB XLOG_SIZECHECK
Argument list: \fIvoid\fP.
Checks the actual file size.
.TP
.SB XLOG_GETFD
Argument list: \fIint *value\fP.
Places in
.I "*value"
the file descriptor of the log file.
.RE
.\" ********************* xlog_write ***********************
.LP
.B xlog_write()
writes a message to the specified xlog. A
.SM NEWLINE
is always appended to the message.
The first occurrence of "%m" in
.I buf
is replaced by a string explaining the current value of
.I errno.
The
.I flags
argument is formed in the same way as in
.B xlog_create().
One additional constant is available:
.RS
.TP 20
.SB XLOG_SET_LEVEL
.I "(XLOG_SYSLOG only)"
the next argument is an integer that should be used as the syslog level
for this message instead of the default level of the xlog.
.RE
.SH "RETURN VALUES"
.B xlog_create()
returns an xlog handle or
.SM NULL
if it fails.
.LP
.B xlog_control()
returns an error code (it returns
.SM XLOG_ENOERROR
if it is successful).
.LP
.B xlog_parms()
returns an error code (it returns
.SM XLOG_ENOERROR
if it is successful).
.SH "SEE ALSO"
openlog(3), syslog(3), closelog(3)
.SH BUGS
.LP
Only the first occurrence of
.I "%m"
is replaced by an errno explanation.
.LP
There is no check for cycles when linking xlogs. In particular it is
possible to link a xlog to itself.


CHANGELOG

This file describes the changes to xinetd. The base version is 2.0.0.

2.0.0: (not released)

2.0.1: (not released)

2.0.2:
	Changes to the Makefile.
	Trivial mods to the usage() function

2.0.3:
	Changes to the Makefile.
	Fixed a bug in the usage() function

2.0.4:
	Distribution versions of the Makefile no longer contain any 
	references to options.opt

2.0.5:
	1) Clarified what an "unlisted RPC service" is in the man page
	2) Fixed a bug in remote_address_check which caused access to be denied
		for all hosts if no_access was set and only_from was not set.
	3) Fixed a bug which caused arbitrary syslog levels if the log_type
		was specified as SYSLOG in the "defaults" entry

2.0.6:
	bug fix in child.c: replaced strx_sprint with strx_print when naming
	interceptor processes


-------------------------------------------------------------------------------

2.1.0: (not released)
	This was mostly a clean-up of 2.0
	List of changes:

	A. The man page has been split into 3 parts:
			xinetd.man			:	man page describing the program
			xinetd.conf.man	:	man page describing the configuration file 
										This file now includes figures about the
										overhead of interception.
			xinetd.log.man		:	man page describing the log file
	B. New service attributes
			rpc_number			: to support unlisted RPC services
			nice					: to set the nice value of forked servers
	C. The IDONLY service flag was added
	D. Now uses the timer library unless NO_TIMERS is defined. If NO_TIMERS
		is not defined, the following are also available:
			a) a new option, -cc, to do periodic consistency checks
			b) timeout for reconfiguration
	E.	Configuration file man page now mentions that access control is 
		based on IP-address instead of domain address.
	F.	The interception code now sets the TCP_NODELAY option.
	G. The timeout when contacting a remote identification server is
		configurable both when the service request has been accepted and 
		when the request is rejected. In the former case it defaults to
		infinity while in the latter it defaults to 30 sec.
	H. The log line ids are now constants in an include file.
	I. The fsma library is no longer used.
	J. Most structure fields were renamed; certain functions were renamed too.
		Access to structure fields is now via macros.
	K. A bug in not restoring the number of descriptors to the soft limit
		was fixed.
	L. The -pid option works
	M. New internal services:
			"servers"		: lists active servers
			"services"		: lists active services

2.1.1:
	xinetd produces more meaningful messages if it can't start logging

2.1.2:
	Fixed 2 bugs:
		1) xinetd would request a port number for unlisted RPC services
			One could get around this by simply specifying the port attribute
			for the service; the specified value would be ignored anyway.
		2) xinetd initialization might fail sometimes in function msg_init()
			syslog logging was specified (i.e. the option -syslog was used)
			The cause of this bug was that the 
				xlog_control( xlog, XLOG_GETFD, &fd ) 
			operation will not fail for xlogs connected to syslog and
			the value of 'fd' would be used in the subsequent fcntl(). That
			value is arbitrary since 'fd' is a local variable.
			If 'fd' did not happen to refer to an open descriptor, the program 
			would terminate since the fcntl() would fail (btw, the success
			of the fcntl() call would be harmless if 'fd' happened to
			refer to an open descriptor).

2.1.3:
	Bug fix:
		only_from/no_access addresses would get inverted on little-endian
		machines when such addresses were specified using the numeric notation
		(for example, 128.138.45.3). This bug was in the numeric_addr function
		which did not convert the result of inet_addr to host-byte-order.
		A work around for the bug would be to use the factorized address 
		notation (for example, 128.138.45.{3})

2.1.4:
	Bug fixes:
		1) in dgram_echo(), sin_len was not being set before the invocation of
			recvfrom
		2) in finger_shutdown(), it was possible for Srdline() to return NULL
			(if the remote end would close the socket without sending anything).
			If the RECORD option was set in the log_on_failure flags, this
			would cause the forked xinetd process which did the recording to
			die since it would try to dereference a NULL pointer.


2.1.5:
	Bug fixes:
		1) in exec_server() service descriptor might be closed when
		   execing server. This made it impossible to start servers
			for 'nowait' services.  The bug occurred only Ultrix version
			4.3a or (probably) later (bug discovered and fix provided
			by doug@seas.smu.edu)
		
		2) for systems that supported supplementary group id's, the
			set_credentials() function did not set those group id's
			(they were being inherited from xinetd). Now initgroups(3)
			is called to set the supplementary group id's properly.
			(bug discovered and fix provided by maf+@osu.edu)

2.1.6:
	Bug fixes:
		1) xinetd will crash after reconfiguration if there is a running
			server for a service that was removed from the configuration and
			which logs on exit.
		2) xinetd forked process falls in infinite loop if identd server
			sends a reply that is missing the ending CR-LF.
			(bug discovered and fix provided by Laurent.Wacrenier@gin.obspm.fr).

	We also change the LOGUSER_SUCCESS_TIMEOUT constant in config.h from
	0 seconds (i.e. infinite timeout) to 30 seconds. This avoids infinite
	waits in case the remote host does not send a RST reply when we attempt
	to connect to the IDENTD port, and there is no server listening at
	that port.


2.1.7:
	Bug fix: the HOST flag in the 'log_on_success' attribute was ignored;
				the code was incorrectly checking if the HOST flag was
				set in the log_on_failure attribute
				(bug discovered by frolich@corrine.cpc.cs.ucf.edu)

2.1.8.1: 
	Added support for TCP redirect to a remote host.
		--bbraun

2.1.8.2: 
	Added support for binding to specific interface.
		--bbraun

2.1.8.3:
	Changed redirect so that a remote port can be specified as well.

libs/src/xlog/xlog.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms 
 * and conditions for redistribution.
 */

#include "config.h"
#include <stdlib.h>
#include <stdarg.h>

#include "xlog.h"
#include "impl.h"

extern struct xlog_ops __xlog_filelog_ops ;
#ifndef NO_SYSLOG
extern struct xlog_ops __xlog_syslog_ops ;
#endif

struct lookup_table
{
	struct xlog_ops	*ops ;
	xlog_e		type ;
} ;

static struct lookup_table ops_lookup_table[] =
{
	{ &__xlog_filelog_ops,	XLOG_FILELOG },
#ifndef NO_SYSLOG
	{ &__xlog_syslog_ops,	XLOG_SYSLOG },
#endif
	{ NULL,                 0           }
} ;


#define CALLBACK( xp, status ) 				\
	if ( (xp)->xl_callback ) 			\
		(*(xp)->xl_callback)( (xlog_h)(xp), status, (xp)->xl_callback_arg )


static void xlog_link( xlog_s *client, xlog_s *server ) ;
static void xlog_unlink( xlog_s *xp ) ;


static struct xlog_ops *xlog_ops_lookup( xlog_e type )
{
	struct lookup_table *ltp ;

	for ( ltp = &ops_lookup_table[ 0 ] ; ltp->ops ; ltp++ )
		if ( ltp->type == type )
			break ;
	return( ltp->ops ) ;
}


/* VARARGS3 */
xlog_h xlog_create( xlog_e type, const char *id, int flags, ... )
{
	xlog_s				*xp ;
	va_list				ap ;
	struct xlog_ops	*xops ;
	int				status ;

	if ( ( xp = NEW( xlog_s ) ) == NULL )
		return( NULL ) ;
	
	if ( id == NULL || ( xp->xl_id = __xlog_new_string( id ) ) == NULL )
	{
		free( xp ) ;
		return( NULL ) ;
	}

	xops = xlog_ops_lookup( type ) ;
	
	if ( xops != NULL )
	{
		va_start( ap, flags ) ;
		xp->xl_ops = xops ;
		status = XL_INIT( xp, ap ) ;
		va_end( ap ) ;

		if ( status == XLOG_ENOERROR )
		{
			xp->xl_flags = flags ;
			xp->xl_type = type ;
			xp->xl_clients = XLOG_NULL ;
			xp->xl_use = XLOG_NULL ;
			return( (xlog_h) xp ) ;
		}
	}

	free( xp->xl_id ) ;
	free( xp ) ;
	return( NULL ) ;
}



static void xlog_link( xlog_s *client, xlog_s *server )
{
	client->xl_use = server ;
	if ( server == NULL )
		return ;

	if ( server->xl_clients == XLOG_NULL )
	{
		INIT_LINKS( client, xl_other_users ) ;
		server->xl_clients = client ;
	}
	else
		LINK( server, client, xl_other_users ) ;
}


static void xlog_unlink( xlog_s *xp )
{
	xlog_s *server = xp->xl_use ;

	/*
	 * Step 1: remove from server chain
	 */
	if ( server != XLOG_NULL )
	{
		if ( server->xl_clients == xp )
			if ( NEXT( xp, xl_other_users ) == xp )
				server->xl_clients = XLOG_NULL ;
			else
				server->xl_clients = NEXT( xp, xl_other_users ) ;
		else
			UNLINK( xp, xl_other_users ) ;
	}

	/*
	 * Step 2: If we have users, clear their link to us.
	 */
	if ( xp->xl_clients != NULL )
	{
		xlog_s *xp2 = xp->xl_clients ;

		do
		{
			xp2->xl_use = XLOG_NULL ;
			xp2 = NEXT( xp2, xl_other_users ) ;
		}
		while ( xp2 != xp->xl_clients ) ;
	}
}


static void xlog_flags( xlog_s *xp, xlog_cmd_e cmd, ... )
{
	va_list ap;
	int	flag;
	int	old_value;
	int	*valp;

	va_start(ap, cmd);
	flag = va_arg( ap, int );
	old_value = ((xp->xl_flags & flag) != 0);
	valp = va_arg( ap, int * );
	va_end(ap);
	if ( cmd == XLOG_SETFLAG ) {
		if ( *valp )
			xp->xl_flags |= flag ;
		else
			xp->xl_flags &= ~flag ;
	}
	*valp = old_value ;
}


void xlog_destroy( xlog_h pxlog )
{
	xlog_s *xp = XP( pxlog ) ;

	xlog_unlink( xp ) ;
	XL_FINI( xp ) ;
	free( xp->xl_id ) ;
	free( xp ) ;
}


/* VARARGS4 */
void xlog_write( xlog_h pxlog, const char buf[], int len, int flags, ... )
{
	xlog_s	*xp = XP( pxlog ) ;
	va_list	ap ;
	int		status ;

	va_start( ap, flags ) ;
	status = XL_WRITE( xp, buf, len, flags, ap ) ;
	va_end( ap ) ;

	if ( status != XLOG_ENOERROR )
	{
		CALLBACK( xp, status ) ;
	}
}


/* VARARGS2 */
int xlog_control( xlog_h pxlog, xlog_cmd_e cmd, ... )
{
	va_list	ap ;
	xlog_s	*xp		= XP( pxlog ) ;
	int		status	= XLOG_ENOERROR ;

	va_start( ap, cmd ) ;

	switch ( cmd )
	{
		case XLOG_LINK:
			xlog_unlink( xp ) ;
			xlog_link( xp, va_arg( ap, xlog_s * ) ) ;
			xp->xl_callback_arg = va_arg( ap, void * ) ;
			break ;
		
		case XLOG_CALLBACK:
			xp->xl_callback = va_arg( ap, voidfunc ) ;
			xp->xl_callback_arg = va_arg( ap, void * );
			break ;
			
		case XLOG_GETFLAG:
		case XLOG_SETFLAG:
			xlog_flags( xp, cmd, ap ) ;
			break ;

		default:
			status = XL_CONTROL( xp, cmd, ap ) ;
	}

	va_end( ap ) ;

	return( status ) ;
}


int xlog_parms( xlog_e type, ... )
{
	va_list	ap ;
	int		status ;

	va_start( ap, type ) ;
	switch ( type )
	{
#ifndef NO_SYSLOG
		case XLOG_SYSLOG:
			status = (*__xlog_syslog_ops.parms)( type, ap ) ;
			break ;
#endif
		case XLOG_FILELOG:
			status = (*__xlog_filelog_ops.parms)( type, ap ) ;
			break ;
		
		default:
			status = XLOG_ENOERROR ;
	}
	va_end( ap ) ;
	return( status ) ;
}

